### PR TITLE
dx: add -protected-branch to pr-auditor

### LINKED ISSUE #33959: Deployment on cloud production that skipped preprod are logged as an exception in github.com/sourcegraph/sec-pr-audit-trail/
Follow-up of the discussion over https://github.com/sourcegraph/sourcegraph/issues/33789. 

In the eventuality of someone deploys directly in the `release` branch, an exception should be logged.

### PR DESCRIPTION
**Pull request by jhchabran**
Fixes #33959 

## Test plan

CI + added unit tests.

### CODE REVIEW

File: dev/pr-auditor/README.md [hunk 1]
```diff
@@ -9,14 +9,17 @@ Learn more: [Testing principles and guidelines](https://docs.sourcegraph.com/dev
 
 This action is primarily designed to run on GitHub Actions, and leverages the [pull request event payloads](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request) extensively.
 
+The `-protected-branch` optional flag enables to open exceptions regardless of the outcome, when a given branch should never be the target of a pull request.
```
Comment (bobheadxi) [inline dev/pr-auditor/README.md:12]: Why is this needed? Shouldn't this be enforced via branch protections?

Comment (jhchabran) [reply dev/pr-auditor/README.md:12]: There are valid cases where one may need to deploy something immediately instead of having to wait for the delay in between prod and preprod: i.e fixing an incident. 

It would put a strong requirement on managing the admin permissions on that repo to select who can do that. This achieves the same thing with more flexibility. 

Does this makes sense for you?

Comment (bobheadxi) [reply dev/pr-auditor/README.md:12]:
```suggestion
The optional `-protected-branch` flag defines a base branch that always opens a PR audit issue to track all pull requests made to it.
```

Comment (jhchabran) [reply dev/pr-auditor/README.md]: Ah right! I understand your confusion now 😊

File: dev/pr-auditor/check.go [hunk 2]
```diff
@@ -14,6 +14,9 @@ type checkResult struct {
 	Reviewed bool
 	// TestPlan is the content provided after the acceptance checklist checkbox.
 	TestPlan string
+	// Protected indicates that the base branch for this PR is protected and merges
+	// are considered to be exceptional and should always be justified.
```
Comment (bobheadxi) [inline dev/pr-auditor/check.go:18]:
```suggestion
	// are considered to be exceptional and should always be logged with a PR audit issue.
```

File: dev/pr-auditor/check.go [hunk 3]
```diff
@@ -23,13 +26,18 @@ func (r checkResult) HasTestPlan() bool {
 }
 
 var (
-	testPlanDividerRegexp       = regexp.MustCompile("(?m)(#+ Test [pP]lan)|(Test [pP]lan:)")
-	noReviewNeededDividerRegexp = regexp.MustCompile("(?m)([nN]o [rR]eview [rR]equired:)")
+	testPlanDividerRegexp       = regexp.MustCompile("(?mi)(#+ test plan)|(test plan:)")
+	noReviewNeededDividerRegexp = regexp.MustCompile("(?mi)(no review required:)")
```
Comment (bobheadxi) [inline dev/pr-auditor/check.go:30]: Can we revert these? We don't really want `TEST PLAN` or `TeSt PlAn`, the specificity of the regex is intentional

Comment (bobheadxi) [reply dev/pr-auditor/check.go:30]: it makes it harder to accidentally match on something wrong as well

File: dev/pr-auditor/check.go [hunk 4]
```diff
@@ -14,6 +14,9 @@ type checkResult struct {
 	Reviewed bool
 	// TestPlan is the content provided after the acceptance checklist checkbox.
 	TestPlan string
+	// Protected indicates that the base branch for this PR is protected and merges
+	// are considered to be exceptional and should always be justified.
+	Protected bool
```
Comment (bobheadxi) [inline dev/pr-auditor/check.go:19]:
```suggestion
	ProtectedBaseBranch bool
```

File: dev/pr-auditor/main.go [hunk 5]
```diff
@@ -167,6 +181,8 @@ func preMergeAudit(ctx context.Context, ghc *github.Client, payload *EventPayloa
 		prState = "failure"
 		stateDescription = "No test plan detected - please provide one!"
 		stateURL = "https://docs.sourcegraph.com/dev/background-information/testing_principles#test-plans"
+	case result.ProtectedBranch:
+
```
Comment (bobheadxi) [inline dev/pr-auditor/main.go:185]:
```suggestion
	case result.ProtectedBranch:
		prState = "success"
		stateDescription = "No action needed, but an exception will be opened post-merge."
```

You can also set a `stateDescription` and `prState`

File: dev/pr-auditor/main.go [hunk 6]
```diff
@@ -58,6 +63,13 @@ func main() {
 	// allowlist of commonly used default branches.
 	case "main", "master", "release":
 		log.Printf("performing checks against allow-listed pull request base %q", ref)
+	case flags.ProtectedBranch:
+		if flags.ProtectedBranch == "" {
+			log.Printf("unknown pull request base %q - discarding\n", ref)
+			return
+		}
```
Comment (bobheadxi) [inline dev/pr-auditor/main.go:70]: This took me a second to wrap my brain around why it's necessary 😆

**Review by bobheadxi — APPROVED** [review]
LGTM!
